src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java

```diff
@@ -1727,4 +1732,77 @@ protected boolean isRunning() {
                 stateMachine.isInState(ReplicationState.IDLE) ||
                 stateMachine.isInState(ReplicationState.OFFLINE);
     }
+
+    protected void waitForPendingFutures() {
+        synchronized (lockWaitForPendingFutures) {
+            if (waitingForPendingFutures) {
+                return;
+            }
+            waitingForPendingFutures = true;
+        }
+
+        Log.v(TAG, "[waitForPendingFutures()] STARTED - thread id: " + Thread.currentThread().getId());
+
+        try {
+            waitForAllTasksCompleted();
+        } catch (Exception e) {
+            Log.e(TAG, "Exception waiting for pending futures: %s", e);
+        }
+
+        // continuous mode, make state IDLE
+        if (isContinuous()) {
+            fireTrigger(ReplicationTrigger.WAITING_FOR_CHANGES);
+        }
+        // one shot mode, make state STOPPING
+        else {
+            triggerStopGraceful();
+        }
+
+        Log.v(TAG, "[waitForPendingFutures()] END - thread id: " + Thread.currentThread().getId());
+
+        synchronized (lockWaitForPendingFutures) {
+            waitingForPendingFutures = false;
+        }
+    }
+
+    protected void waitForAllTasksCompleted() {
+        // NOTE: Wait till all queue becomes empty
+        while ((batcher != null && batcher.count() > 0) ||
+                (pendingFutures != null && pendingFutures.size() > 0)) {
+            // Wait for batcher (inbox) completed
+            waitBatcherCompleted(batcher);
+
+            // wait for pending featurs completed
+            waitPendingFuturesCompleted(pendingFutures);
+        }
+    }
+
+    protected static void waitBatcherCompleted(Batcher<RevisionInternal> b) {
+        // Wait for batcher completed
+        if (b != null) {
+            // if batcher delays task execution, need to wait same amount of time. (0.5 sec or 0 sec)
+            try {
+                Thread.sleep(b.getDelay());
```

Comment:
@hideki I couldn't think of the reason why the sleep here is needed. Batcher.waitForpendingFutures() will wait until all tasks in the queues are done.